### ISSUE TITLE
Elasticsearch 2.3.1 accept a field both as a string and as an object on the same index

### ISSUE THREAD

**Issue by segalziv** [elastic/elasticsearch#18450]
As I understand, it should be impossible to use the same field name with more then one field type on the same index, regardless of the document type. 

But, on 2.3.1 we sometimes see that the index accepts a field with both string and object types (easy to reproduce - see bellow).

After this point the behaviour is not that predictive as some docs gets in and other don't with IllegalArgumentException (see log samples bellow).

Can you please confirm this is a bug or make it clear when a field can be used with more then one field-type?

Thanks!

**Elasticsearch version**: 2.3.1

**JVM version**: 1.7.0_95

**OS version**: Linux AMD64 3.13.0-48-generic

**Description of the problem including expected versus actual behavior**:

**Steps to reproduce**:

1. POST es-host:9200/index-name/type1  { "field-name" : "a string value" }

2. POST es-host:9200/index-name/type2  
{ "field-name" : { "inner-field" : "now an object"} }

3. GET es-host:9200/index-name/_mappings 
```
{
    "index-name": {
        "mappings": {
            "type1": {
                "properties": {
                    "field-name": {
                        "type": "string"
                    }
                }
            },
            "type2": {
                "properties": {
                    "field-name": {
                        "properties": {
                            "inner-field": {
                                "type": "string"
                            }
                        }
                    }
                }
            }
        }
    }
}
```
4. On this point indexing to the index fails on some cases - either when trying to use the same field ('field-name') on new types and on the existing types and even (hard to reproduce) on docs without this specific field.

**Provide logs **:

From the master node:

```
[2016-05-11 08:39:18,659][DEBUG][action.admin.indices.mapping.put] [master-xxx] failed to put mappings on indices [[index1]], type [client]
java.lang.IllegalArgumentException: Field [os] is defined as a field in mapping [elb] but this name is already used for an object in other types
        at org.elasticsearch.index.mapper.MapperService.checkObjectsCompatibility(MapperService.java:473)
        at org.elasticsearch.index.mapper.MapperService.merge(MapperService.java:336)
        at org.elasticsearch.index.mapper.MapperService.merge(MapperService.java:289)
        at org.elasticsearch.cluster.metadata.MetaDataMappingService$PutMappingExecutor.execute(MetaDataMappingService.java:223)
        at org.elasticsearch.cluster.service.InternalClusterService.runTasksForExecutor(InternalClusterService.java:468)
        at org.elasticsearch.cluster.service.InternalClusterService$UpdateTask.run(InternalClusterService.java:772)
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:231)
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:194)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)

From the data node:

[2016-05-11 08:42:58,381][DEBUG][action.bulk              ] [PROD-data-elasticsearch2-4-i-9cbda518] [index1][0] failed to execute bulk item (index) index {[index1][iis][AVSe-7uvn-xxxxxxxxx], source[{"cs-method":"GET","cs-uri-stem":"/deploy/General/CheckHealth","cs-uri-query":"-","sc-substatus":0,"type":"iis","cs-username":"-","os_iis":"Other","host":"IP-xxxx","timestamp":"2016-05-11 08:41:56","app":"player","s-sitename":"0.0.0.0","message":"2016-05-11 08:41:56 0.0.0.0 GET /deploy/General/CheckHealth - 80 - 1.1.1.1 ELB-HealthChecker/1.0 - 200 0 0 0","env":"dev","time-taken":0,"cs":["ELB-HealthChecker/1.0","-"],"@timestamp":"2016-05-11T08:41:56.000Z","response":200,"s-port":"80","name":"Other","os_name":"Other","c-ip":"1.1.1.1","device":"Other"}]}
java.lang.IllegalArgumentException: Field [os] is defined as a field in mapping [elb] but this name is already used for an object in other types
        at org.elasticsearch.index.mapper.MapperService.checkObjectsCompatibility(MapperService.java:473)
        at org.elasticsearch.index.mapper.MapperService.merge(MapperService.java:336)
        at org.elasticsearch.index.mapper.MapperService.merge(MapperService.java:289)
        at org.elasticsearch.cluster.metadata.MetaDataMappingService$PutMappingExecutor.execute(MetaDataMappingService.java:223)
        at org.elasticsearch.cluster.service.InternalClusterService.runTasksForExecutor(InternalClusterService.java:468)
        at org.elasticsearch.cluster.service.InternalClusterService$UpdateTask.run(InternalClusterService.java:772)
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:231)
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:194)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by clintongormley** [comment]
This is indeed a bug, and is already fixed in master.  @jpountz @rjernst is this fix backportable?

**Comment by segalziv** [comment]
ok. thanks for the fast response!

I really hope this is indeed going to get back ported.

**Comment by rjernst** [comment]
@clintongormley I think you mean #17568, which was already backported and released in 2.3.2?

**Comment by clintongormley** [comment]
Ahhhh thanks @rjernst - I missed that.  Tried the above recreation in 2.3.2 and it fails (as it should).

thanks.  Fixed by #17568

**Comment by segalziv** [comment]
Thanks guys!
